src/gdbstub_impl/ext/host_io.rs

```diff
@@ -1,7 +1,21 @@
 use super::prelude::*;
 use crate::arch::Arch;
 use crate::protocol::commands::ext::HostIo;
-use crate::target::ext::host_io::{HostStat, PreadOutput};
+use crate::target::ext::host_io::{HostIoError, HostStat, PreadOutput};
+use crate::GdbStubError;
+
+macro_rules! handle_hostio_result {
+    ( $ret:ident, $res:ident, $callback:expr) => {{
+        match $ret {
+            Ok(fd) => $callback(fd)?,
+            Err(HostIoError::Errno(errno)) => {
+                $res.write_str("F-1,")?;
+                $res.write_num(errno as i32)?;
+            }
+            Err(HostIoError::Fatal(e)) => return Err(GdbStubError::TargetError(e)),
+        }
+    }};
+}
```

Comment:
A few suggestions on how to improve this macro (from someone who's written some [pretty gnarly](https://github.com/daniel5151/analog_literals) Rust macros in the past):

- You can define this macro as part of `handle_host_io`'s body, which would save you from having to explicitly bind `$res`, as it would already be in scope as `res`.
- Instead of using a callback, you should use a `$callback:block` instead.
- Use the `expr` fragment-specifier instead of `ident` to avoid having to introduce an extra `result` binding
- With these changes in place, tweak the syntax to look a bit more "familiar"
    - This is pretty subjective, but I like it when macros embed a Rust-y looking syntax into them
- Use the `pat` fragment-specifier instead of `ident` for `$fd` to enable inline destructuring (not that it really matters here, but hey, good to know)

The final macro would look something like:

```rust
macro_rules! handle_hostio_result {
    (
        if let Ok($fd:pat) = $ret:expr => $callback:block
    ) => {{
        match $ret {
            Ok($fd) => $callback,
            Err(HostIoError::Errno(errno)) => {
                res.write_str("F-1,")?;
                res.write_num(errno as i32)?;
            }
            Err(HostIoError::Fatal(e)) => return Err(GdbStubError::TargetError(e)),
        }
    }};
}
```

and would be invoked as

```rust
handle_hostio_result! {
    if let Ok(fd) = ops.open(cmd.filename, cmd.flags, cmd.mode) => {
        // ...
    }
}
```

Reply:
Awesome!